### PR TITLE
Add a script to manipulate the poetry setup.

### PR DESCRIPTION
**Pull request by jesnie**
Add a script (`poetryenv`) that manipulates the files `poetry` depends on, to allow us to run `poetry` with different versions of Python and dependencies.

### CODE REVIEW

File: poetryenv [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+    write_toml(pyproject, _PYPROJECT_TOML)
+
+
+def poetryenv(env_dir: Path, poetry_args: List[str]) -> int:
+    assert _PYPROJECT_TOML.is_file()
+    assert env_dir.is_dir()
+    _BACKUP_DIR.mkdir()
+
+    try:
+        with shuffle(_PYPROJECT_TOML, env_dir, derived=False), shuffle(
```
Comment (uri-granta) [inline poetryenv:119]: (I think you can put the three shuffes all in parentheses, which should make them align better)

File: poetryenv [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+            _POETRY_LOCK, env_dir, derived=True
+        ), shuffle(_VENV_DIR, env_dir, derived=True):
+            env_pyproject_overrides_toml = env_dir / "pyproject_overrides.toml"
+            update_pyproject_toml(env_pyproject_overrides_toml)
+
+            command = " ".join(["poetry"] + poetry_args)
+            return subprocess.run(command, shell=True, check=False).returncode
```
Comment (uri-granta) [inline poetryenv:126]: Does the following not work?

```suggestion
            return subprocess.run("poetry", *poetry_args, shell=True, check=False).returncode
```

**Review by uri-granta — APPROVED** [review]
Surprisingly tidy, well done!

File: poetryenv [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+
+def override_toml(root: tomlkit.TOMLDocument, overrides: tomlkit.TOMLDocument) -> None:
+    for key, override_value in overrides.items():
+        if (key in root) and (hasattr(root[key], "items")):
+            override_toml(root[key], override_value)
+        else:
+            root[key] = override_value
+
+
+def update_pyproject_toml(overrides_path: Path) -> None:
```
[the remark this answers is not in the document]
Comment (jesnie) [reply poetryenv]: Sure.

File: poetryenv (hunk 1 above)
Comment (jesnie) [reply poetryenv]: > (I think you can put the three shuffes all in parentheses, which should make them align better)

It doesn't seem so. I can use a `contextlib.ExitStack` to make it a bit more readable, though.

File: poetryenv (hunk 2 above)
Comment (jesnie) [reply poetryenv]: No, that doesn't seem to work.

File: poetryenv [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+            if root.exists():
+                root.replace(env_path)
+            if backup_path.exists():
+                backup_path.replace(root)
+        else:
+            backup_path.replace(root)
+
+
+def read_toml(path: Path) -> tomlkit.TOMLDocument:
+    with open(path, "rt", encoding="utf-8") as f:
```
[the remark this answers is not in the document]
Comment (jesnie) [reply poetryenv:98]: :+1:

**Pull request merged**